File: app/src/main/java/com/yunfie/illustia/settings/db/IllustiaDatabase.java
package com.yunfie.illustia.settings.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

@Database(
        entities = {
                SearchHistoryEntity.class,
                FavoriteTagEntity.class,
                ViewHistoryEntity.class,
                AccountEntity.class,
                SavedIllustEntity.class,
                SavedIllustPageEntity.class
        },
        version = 2,
        exportSchema = false
)
public abstract class IllustiaDatabase extends RoomDatabase {
    public abstract SettingsDao settingsDao();

    private static volatile IllustiaDatabase INSTANCE;

    public static IllustiaDatabase getInstance(Context context) {
        IllustiaDatabase current = INSTANCE;
        if (current != null) {
            return current;
        }
        synchronized (IllustiaDatabase.class) {
            current = INSTANCE;
            if (current == null) {
                current = Room.databaseBuilder(
                                context.getApplicationContext(),
                                IllustiaDatabase.class,
                                "illustia.db"
                        )
                        .fallbackToDestructiveMigration()
                        .build();
                INSTANCE = current;
            }
            return current;
        }
    }
}

File: shared/src/androidMain/kotlin/com/yunfie/illustia/settings/db/IllustiaDatabase.kt
package com.yunfie.illustia.settings.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

@Database(
    entities = [
        SearchHistoryEntity::class,
        FavoriteTagEntity::class,
        ViewHistoryEntity::class,
        AccountEntity::class,
        SavedIllustEntity::class,
        SavedIllustPageEntity::class,
    ],
    version = 3,
    exportSchema = false,
)
abstract class IllustiaDatabase : RoomDatabase() {
    abstract fun settingsDao(): SettingsDao

    companion object {
        @Volatile
        private var INSTANCE: IllustiaDatabase? = null

        private val MIGRATION_2_3 =
            object : Migration(2, 3) {
                override fun migrate(db: SupportSQLiteDatabase) {
                    db.execSQL("ALTER TABLE saved_illusts ADD COLUMN xRestrict INTEGER NOT NULL DEFAULT 0")
                }
            }

        fun getInstance(context: Context): IllustiaDatabase =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: Room.databaseBuilder(
                    context.applicationContext,
                    IllustiaDatabase::class.java,
                    "illustia.db",
                )
                    .addMigrations(MIGRATION_2_3)
                    .fallbackToDestructiveMigration()
                    .build()
                    .also { INSTANCE = it }
            }
    }
}
